sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/GpuHyperLogLogPlusPlus.scala

```diff
@@ -0,0 +1,196 @@
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (input: cudf.ColumnVector) => {
+      if (input.getNullCount == input.getRowCount) {
+        // For NullType column or all values are null,
+        // return a struct scalar: struct(0L, 0L, ..., 0L)
+        val numCols = (1 << precision) / 10 + 1
+        withResource(cudf.ColumnVector.fromLongs(0L)) { cv =>
+          // Underlying uses deep-copy, so we can reuse this `cv` and fill multiple times.
+          val cvs: Array[cudf.ColumnView] = Array.fill(numCols)(cv)
+          cudf.Scalar.structFromColumnViews(cvs: _*)
+        }
+      } else {
+        val hll = new HyperLogLogPlusPlusHostUDF(AggregationType.Reduction, precision)
+        input.reduce(ReductionAggregation.hostUDF(hll), DType.STRUCT)
+      }
+    }
+  override lazy val groupByAggregate: GroupByAggregation = {
+    val hll = new HyperLogLogPlusPlusHostUDF(AggregationType.GroupBy, precision)
+    GroupByAggregation.hostUDF(hll)
+  }
+  override val name: String = "CudfHyperLogLogPlusPlus"
+}
+
+case class CudfMergeHLLPP(override val dataType: DataType,
+    precision: Int)
+  extends CudfAggregate {
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (input: cudf.ColumnVector) => {
+      val hll = new HyperLogLogPlusPlusHostUDF(AggregationType.ReductionMerge, precision)
+      input.reduce(ReductionAggregation.hostUDF(hll), DType.STRUCT)
+    }
+
+  override lazy val groupByAggregate: GroupByAggregation = {
+    val hll = new HyperLogLogPlusPlusHostUDF(AggregationType.GroupByMerge, precision)
+    GroupByAggregation.hostUDF(hll)
+  }
+
+  override val name: String = "CudfMergeHyperLogLogPlusPlus"
+}
+
+/**
+ * Perform the final evaluation step to compute approximate count distinct from sketches.
+ * Input is long columns, first construct struct of long then feed to cuDF
+ */
+case class GpuHyperLogLogPlusPlusEvaluation(childExpr: Expression,
+    precision: Int)
+  extends GpuExpression with ShimExpression {
+  override def dataType: DataType = LongType
+
+  override def nullable: Boolean = false
+
+  override def prettyName: String = "HyperLogLogPlusPlus_evaluation"
+
+  override def children: Seq[Expression] = Seq(childExpr)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    withResource(childExpr.columnarEval(batch)) { sketches =>
+      val distinctValues = HyperLogLogPlusPlusHostUDF.estimateDistinctValueFromSketches(
+        sketches.getBase, precision)
+      GpuColumnVector.from(distinctValues, LongType)
+    }
+  }
+}
+
+/**
+ * Gpu version of HyperLogLogPlusPlus
+ * Spark APPROX_COUNT_DISTINCT on NULLs returns zero
+ */
+case class GpuHyperLogLogPlusPlus(childExpr: Expression, relativeSD: Double)
+  extends GpuAggregateFunction with Serializable {
+
+  // Consistent with Spark
+  private lazy val precision: Int = GpuHyperLogLogPlusPlus.computePrecision(relativeSD)
+
+  private lazy val numRegistersPerSketch: Int = 1 << precision;
```

Comment:
NIT:
Seems used only once for the "numWords" computation. No need to be a member field.